File: BluetoothLightLib/src/main/java/com/telink/bluetooth/Command.java
/*
 * Copyright (C) 2015 The Telink Bluetooth Light Project
 *
 */
package com.telink.bluetooth;

import com.telink.util.Arrays;

import java.util.UUID;

public class Command {

    public UUID serviceUUID;
    public UUID characteristicUUID;
    public CommandType type;
    public byte[] data;
    public Object tag;
    public int delay;

    public Command() {
        this(null, null, CommandType.WRITE);
    }

    public Command(UUID serviceUUID, UUID characteristicUUID, CommandType type) {
        this(serviceUUID, characteristicUUID, type, null);
    }

    public Command(UUID serviceUUID, UUID characteristicUUID, CommandType type,
                   byte[] data) {
        this(serviceUUID, characteristicUUID, type, data, null);
    }

    public Command(UUID serviceUUID, UUID characteristicUUID, CommandType type,
                   byte[] data, Object tag) {

        this.serviceUUID = serviceUUID;
        this.characteristicUUID = characteristicUUID;
        this.type = type;
        this.data = data;
        this.tag = tag;
    }

    public static Command newInstance() {
        return new Command();
    }

    public void clear() {
        this.serviceUUID = null;
        this.characteristicUUID = null;
        this.data = null;
    }

    @Override
    public String toString() {
        String d = "";

        if (data != null)
            d = Arrays.bytesToHexString(this.data, ",");

        return "{ tag : " + this.tag + ", type : " + this.type
                + " characteristicUUID :" + characteristicUUID.toString() + " data: " + d + " delay :" + delay + "}";
    }

    public enum CommandType {
        READ, WRITE, WRITE_NO_RESPONSE, ENABLE_NOTIFY, DISABLE_NOTIFY
    }

    public interface Callback {

        void success(Peripheral peripheral, Command command, Object obj);

        void error(Peripheral peripheral, Command command, String errorMsg);

        boolean timeout(Peripheral peripheral, Command command);
    }
}

File: BluetoothLightLib/src/main/java/com/telink/light/Command.kt
package com.telink.light

import com.telink.util.Arrays
import java.util.*

class Command @JvmOverloads constructor(var serviceUUID: UUID? = null, var characteristicUUID: UUID? = null, var type: CommandType = CommandType.WRITE,
                                        var data: ByteArray? = null, var tag: Any? = null) {
    var delay = 0

    fun clear() {
        serviceUUID = null
        characteristicUUID = null
        data = null
    }

    override fun toString(): String {
        var d = ""
        if (data != null)
            d = Arrays.bytesToHexString(data, ",")

        return ("{ tag : $tag, type : $type , characteristicUUID :$characteristicUUID ,data: $d . delay :$delay}")
    }

    enum class CommandType {
        READ, WRITE, WRITE_NO_RESPONSE, ENABLE_NOTIFY, DISABLE_NOTIFY
    }

    interface Callback {

        fun success(peripheral: Peripheral, command: Command, obj: Any)

        fun error(peripheral: Peripheral, command: Command, errorMsg: String)

        fun timeout(peripheral: Peripheral, command: Command): Boolean
    }

    companion object {
        fun newInstance() = Command()
    }
}
